include/cetl/span.h

```diff
@@ -1 +1,772 @@
+    ///                 conversion](https://en.cppreference.com/w/cpp/language/implicit_conversion#Qualification_conversions).
+    /// @param source The span to copy from. The resulting span has `size() == source.size()` and `data() ==
+    ///               source.data()`
+    /// @see std::span(std::span&)
+    template <typename DeducedElementType,
+              typename std::enable_if<std::is_convertible<DeducedElementType (*)[], element_type (*)[]>::value,
+                                      bool>::type = true>
+    explicit constexpr span(const span<DeducedElementType, cetl::dynamic_extent>& source) noexcept
+        : data_(source.data())
+    {
+        CETL_DEBUG_ASSERT(extent == source.size(),
+                          "Undefined behavior: providing a dynamic span with a size different from this static "
+                          "span's.");
+    }
+
+    /// Copy constructor to create a span from another span.
+    /// @tparam DeducedElementType  The element type of the source span.
+    /// @tparam type                Enables this override only if the element conversion is a simple [qualification
+    ///                 conversion](https://en.cppreference.com/w/cpp/language/implicit_conversion#Qualification_conversions).
+    /// @param source The span to copy from. The resulting span has `size() == source.size()` and
+    ///               `data() == source.data()`. For this overload `extent == source::extent` is also true and
+    ///               `extent != cetl::dynamic_extent`.
+    /// @see std::span()
+    template <typename DeducedElementType,
+              typename std::enable_if<std::is_convertible<DeducedElementType (*)[], element_type (*)[]>::value,
+                                      bool>::type = true>
+    constexpr span(const span<DeducedElementType, Extent>& source) noexcept
+        : data_(source.data())
+    {
+    }
+
+    /// Default copy constructor
+    /// @see std::span::span()
+    constexpr span(const span&) noexcept = default;
+
+    ///@}
+    // +----------------------------------------------------------------------+
+    ///@{ @name Operators
+    // +----------------------------------------------------------------------+
+    /// Default assignment operator.
+    /// @param rhs    Right-hand-side of the assignment expression.
+    /// @return reference to `*this`
+    /// @see std::span::operator=()
+    constexpr span& operator=(const span& rhs) noexcept = default;
+
+    ///
+    /// Reference to an element in the span.
+    /// @param idx  The 0-based index in the span. The behavior of this method is
+    ///             undefined if idx is >= to size.
+    /// @return A reference to an element.
+    /// @see std::span::operator[]()
+    constexpr reference operator[](size_type idx) const
+    {
+        CETL_DEBUG_ASSERT((idx < Extent), "Indexing outside of the span is undefined.");
```

Comment:
This is about all you can do for static spans except wrapping the index. Can this span be empty? (nullptr, zero size?)

Reply:
Yes. If you define the Extent as 0 the data_ pointer is null. I should expand this to be: `(Extent > 0) && (idx < Extent)`